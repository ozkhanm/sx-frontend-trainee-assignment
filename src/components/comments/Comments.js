import React, {useEffect} from "react";
import {connect} from "react-redux";

import Comment from "../comment/Comment";

import {ActionCreator} from "../../reducer/action-creator";
import {Operation} from "../../reducer/reducer";

const Comments = (props) => {
    const {articleComments, activeArticle, isCommentLoaded, 
        getActiveArticle, changeRefreshStatus, changeCommentsLoadingStatus} = props;

    useEffect(() => {
        const refreshInterval = setInterval(() => {
            changeRefreshStatus(true);
            changeCommentsLoadingStatus(false);
            getActiveArticle(activeArticle.id);
          }, 60000);

        return () => {
            clearInterval(refreshInterval);
        };
    });

    const getCommentsElements = (comments) => {
        return comments.map((it) => <Comment key={it.id} comment={it}/>);
    };

    return (
        <div className="comment-block">
            {isCommentLoaded ? getCommentsElements(articleComments) : ``}
        </div>
    );
};

const mapStateToProps = (state) => ({
    articleComments: state.articleComments,
    isCommentLoaded: state.isCommentLoaded,
    activeArticle: state.activeArticle
});

const mapDispatchToProps = (dispatch) => ({
    getActiveArticle(id) {
        dispatch(Operation.getActiveArticle(id));
    },
    changeRefreshStatus(status) {
        dispatch(ActionCreator.changeRefreshStatus(status));
    },
    changeCommentsLoadingStatus(status) {
        dispatch(ActionCreator.changeCommentsLoadingStatus(status));
    },
});

export {Comments};
export default connect(mapStateToProps, mapDispatchToProps)(Comments);