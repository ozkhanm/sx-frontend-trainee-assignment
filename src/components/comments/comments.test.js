import React from "react";
import renderer from "react-test-renderer";

import {Comments} from "./Comments";

it(`Comments container renders correctly if data is fetching`, () => {
    const tree = renderer
        .create(<Comments articleComments={[]} isCommentLoaded={false}/>)
        .toJSON();
  
    expect(tree).toMatchSnapshot();
});
  
it(`Comments container renders correctly if data fetched`, () => {
    const articleComments = [
        {
            by: "georgeburdell",
            id: 30387999,
            parent: 30386645,
            kids: [
                {
                    by: "CitizenKane",
                    id: 30390670,
                    parent: 30386645,
                    text: "Hey Ian, thanks for all the very thoughtful content over the years. I started reading Anandtech shortly before Anand left and I remember being worried. However, in hindsight there was clearly nothing to worry about.<p>Your analysis helped convince me to go all in on some AMD stock purchases that ended up really saving me over past year. It’s beyond appreciated, and I’m far more knowledgeable about semiconductors thanks to you! Best of luck in your new journey",
                    time: 1645216278,
                    type: "comment"
                }
            ],
            text: "Not unexpected.  He’s been building his own brand (TechTechPotato) on YouTube for a while.",
            time: 1645202735,
            type: "comment"
        },
        {
            by: "Katydid",
            id: 30389216,
            parent: 30386645,
            text: "Cheers to the best in the business, here&#x27;s wishing Ian incredible success. Thanks for so much great reporting and analysis over the years. Few understand how challenging such a job is, having your daily work product (which has to balance technical depth and reader appeal) on stage for the world to grill, year in and out. Ian nailed it with depth, energy, and humor. Can&#x27;t wait to see what&#x27;s next from the good Dr. Cutress. - N",
            time: 1645208546,
            type: "comment"
        }
    ];
    const tree = renderer
        .create(<Comments articleComments={articleComments} isCommentLoaded={true}/>)
        .toJSON();
  
    expect(tree).toMatchSnapshot();
});