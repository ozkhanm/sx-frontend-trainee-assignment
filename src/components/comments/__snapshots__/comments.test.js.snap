// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments container renders correctly if data fetched 1`] = `
<div
  className="comment-block"
>
  <div
    className="comment-item-block"
  >
    <div
      className="comment-parent-block"
    >
      <div
        className="comment-item-header-block"
      >
        <p
          className="comment-item"
        >
          by: 
          georgeburdell
        </p>
        <p
          className="comment-item"
        >
          at: 
          18.02.2022, 19:45
        </p>
      </div>
      <p
        className="comment-item-text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Not unexpected.  Heâ€™s been building his own brand (TechTechPotato) on YouTube for a while.",
          }
        }
      />
    </div>
    <button
      className="comment-show-more-block"
      onClick={[Function]}
      type="button"
    >
      show more
    </button>
  </div>
  <div
    className="comment-item-block"
  >
    <div
      className="comment-parent-block"
    >
      <div
        className="comment-item-header-block"
      >
        <p
          className="comment-item"
        >
          by: 
          Katydid
        </p>
        <p
          className="comment-item"
        >
          at: 
          18.02.2022, 21:22
        </p>
      </div>
      <p
        className="comment-item-text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Cheers to the best in the business, here&#x27;s wishing Ian incredible success. Thanks for so much great reporting and analysis over the years. Few understand how challenging such a job is, having your daily work product (which has to balance technical depth and reader appeal) on stage for the world to grill, year in and out. Ian nailed it with depth, energy, and humor. Can&#x27;t wait to see what&#x27;s next from the good Dr. Cutress. - N",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Comments container renders correctly if data is fetching 1`] = `
<div
  className="comment-block"
>
  
</div>
`;
